Changes requested 2022-08-12

haishulud opened the pull request:
- One main job is divided into two jobs: test and build
- Workflow time is calculated for both jobs
- SHA1 is used to tag Docker images
- Applies some of the best practices for GitHub Actions
- Rids of unnecessary 3rd party Actions for better security
- Notifications now work for PR (as well as for PUSH) and sent to committer email)

haishulud committed:
sha edit #43346

haishulud committed:
ds

haishulud committed:
gfh

haishulud committed:
email

haishulud committed:
889

haishulud committed:
2022-08-14

haishulud committed:
2022-08-14

haishulud committed:
2022-08-14 cleanup

haishulud committed:
2022-08-12

haishulud committed:
Changes requested 2022-08-12

- One job divided into two jobs: test and build
- Workflow time is calculated for both jobs
- SHA1 is used to tag Docker images
- Applies some of the best practices for GitHub Actions
- Rids of unnecessary 3rd party Actions for better security
- Notifications now work for PR (sent to committer email)

haishulud committed:
2022-08-12

haishulud committed:
Changes requested 2022-08-12

- One main job is divided into two jobs: test and build
- Workflow time is calculated for both jobs
- SHA1 is used to tag Docker images
- Applies some of the best practices for GitHub Actions
- Rids of unnecessary 3rd party Actions for better security
- Notifications now work for PR (sent to committer email)

haishulud committed:
2022-08-15

haishulud committed:
Merge branch 'master' into dev

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -3,25 +3,23 @@ name: myCI
-      contents: read  #permissions for publishing Docker images to Github Packages
-      packages: write #permissions for publishing Docker images to Github Packages
-          
+  test_job:
```
make shorter semantic names
like `test` is enough no need to append `_job` etc.

haishulud replied:
ok

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -3,25 +3,23 @@ name: myCI
-    #branches: master
+    branches: [master, dev]
+  pull_request:
+    branches: [master, dev]
```
so it is dangerous to run this pipeline. dev/master should be separate pipelines

haishulud replied:
done

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -33,18 +31,10 @@ jobs:
       - name: Install dependencies
```
use shorter semantic names for steps. `install` is enough. always prefer lowercase
`Linting -> lint`. Use verbs. like you step should answer "I will XXX", i.e. i will install, I will lint, I will test, I will build etc.

haishulud replied:
ok, done

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -33,18 +31,10 @@ jobs:
+      - name: Linting
+        run: npm run lint
+     
       - name: Test phase
```
like I would call it `run`

dmitry-mightydevops replied:
as we still belong to the job `test`

haishulud replied:
done

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
       #      https://api.github.com/repos/${{ github.actor }}/${{ github.repository }}/actions/runs/${{ github.run_id }}/timing
 
+
+  build_job:
```
build

haishulud replied:
ok

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+
+      - name: build Docker image on push event
+        if: github.event_name == 'push'
+        run: docker build -t $REGISTRY/$IMAGE_NAME:${GITHUB_SHA::7} -t $REGISTRY/$IMAGE_NAME:latest .
```
- let's tag with first 7 symbols of SHA1 of the commit ID. 
- and also tag it with `latest`
then push both tags to remote registry

haishulud replied:
done

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+        run: docker build -t $REGISTRY/$IMAGE_NAME:${GITHUB_SHA::7} -t $REGISTRY/$IMAGE_NAME:latest .
+
+      - name: build Docker image on PR event
+        if: github.event_name == 'pull_request'
```
why do you distinguish between the push and pr events for the build? 
you don't tag it differently - so it just code duplication. same issue with few steps below (push)

haishulud replied:
I've got the wrong(?) idea that commit hash is extracted with different commands on push and PR.
Also official documentation says (https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request):
"Note that GITHUB_SHA for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use github.event.pull_request.head.sha instead."
Due to my lack of experience I'm not sure whether I need 'the last merge commit' or 'the last commit to the head branch'.

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+          PR_SHA7=${{github.event.pull_request.head.sha}}
+          docker build -t $REGISTRY/$IMAGE_NAME:${PR_SHA7::7} -t $REGISTRY/$IMAGE_NAME:latest .
+
+      - name: push Docker image on push event
+        if: github.event_name == 'push'
+        run: docker push $REGISTRY/$IMAGE_NAME:${GITHUB_SHA::7} && docker push $REGISTRY/$IMAGE_NAME:latest
+
+      - name: push Docker image on PR event
+        if: github.event_name == 'pull_request'
+        run: |
+          PR_SHA7=${{github.event.pull_request.head.sha}}
+          docker push $REGISTRY/$IMAGE_NAME:${PR_SHA7::7} && docker push $REGISTRY/$IMAGE_NAME:latest
```
why do you distinguish between the push and pr events for the push? 
you don't tag it differently - so it just code duplication. same issue with few steps below (push)

haishulud replied:
I've got the wrong(?) idea that commit hash is extracted with different commands on push and PR.
Also official documentation says (https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request):
"Note that GITHUB_SHA for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use github.event.pull_request.head.sha instead."
Due to my lack of experience I'm not sure whether I need 'the last merge commit' or 'the last commit to the head branch'.

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+      - name: build Docker image on PR event
+        if: github.event_name == 'pull_request'
+        run: |
+          PR_SHA7=${{github.event.pull_request.head.sha}}
```
introduce this as env variable and reuse. or just reference `${GITHUB_SHA::7}`

haishulud replied:
referenced ${GITHUB_SHA::7}

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+          PR_SHA7=${{github.event.pull_request.head.sha}}
+          docker push $REGISTRY/$IMAGE_NAME:${PR_SHA7::7} && docker push $REGISTRY/$IMAGE_NAME:latest
+
+      #---Build Docker image, Login & Publish to Docker Packages - WITH 3RD PARTY ACTIONS---
+      #- name: Log in to the Container registry
+      #  uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
+      #  with:
+      #    registry: $REGISTRY
+      #    username: ${{ github.actor }} #${{ github.repository_owner }}
+      #    password: ${{ secrets.GITHUB_TOKEN }}
       
+      #- name: Extract metadata (tags, labels) for Docker
+      #  id: meta
+      #  uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
+      #  with:
+      #    images: $REGISTRY/$IMAGE_NAME
+            
+      #- name: Build and push Docker image
+      #  uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
+      #  with:
+      #    context: ./angular
+      #    push: true
+      #    tags: ${{ steps.meta.outputs.tags }}
+      #    labels: ${{ steps.meta.outputs.labels }}
```
don't keep the commented out code in the repo.

haishulud replied:
ok

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
         run: |
-          echo "::set-output name=starttime::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].started_at')" && echo "::set-output name=endtime::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '. ...
-      
+          echo "::set-output name=starttime_test::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].started_at')" && echo "::set-output name=endtime_test::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jo ...
+          echo "::set-output name=starttime_build::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[1].started_at')" && echo "::set-output name=endtime_build::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/ ...
```
1. that code is very hard to read. Pls improve readability
2. what if you decide to break the pipeline into more jobs? Your finish time calc would be incorrect.

haishulud replied:
changed time calculation approach

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
-          completed_at=${{steps.execution_timestamps.outputs.endtime}}
+          started_at_test=${{steps.execution_timestamps.outputs.starttime_test}}
+          completed_at_test=${{steps.execution_timestamps.outputs.endtime_test}}
+          started_at_build=${{steps.execution_timestamps.outputs.starttime_build}}
+          completed_at_build=${{steps.execution_timestamps.outputs.endtime_build}}
 
-          START_TIME=$(date -d "$started_at" +%s)
-          END_TIME=$(date -d "$completed_at" +%s)
+          START_TIME_TEST=$(date -d "$started_at_test" +%s)
+          END_TIME_TEST=$(date -d "$completed_at_test" +%s)
+          START_TIME_BUILD=$(date -d "$started_at_build" +%s)
+          END_TIME_BUILD=$(date -d "$completed_at_build" +%s)
 
-          ELAPSE=$(( $END_TIME - $START_TIME ))
+          ELAPSE_TEST=$(( $END_TIME_TEST - $START_TIME_TEST ))
+          ELAPSE_BUILD=$(( $END_TIME_BUILD - $START_TIME_BUILD ))
+          ELAPSE=$(( $ELAPSE_TEST + $ELAPSE_BUILD ))
           echo $ELAPSE
           echo "$(($ELAPSE/60/60))h $(($ELAPSE/60%60))m $(($ELAPSE%60))s"
           echo "::set-output name=elpsd::$ELAPSE"
           echo "::set-output name=el_hours::$(($ELAPSE/60/60))"
           echo "::set-output name=el_minutes::$(($ELAPSE/60%60))"
```
that is too much implementation for the duration calculation. Like you went to your local grocery store by flying to Africa and back first.

let's find a better way.

haishulud replied:
found a better way: dedicated steps at the beginning and the end of the workflow to get timestamps.

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
+      - name: get email for Committer on PR or pusher on PUSH
+        id: email
+        run: |
+          if [[ ${{ github.event_name }} == 'pull_request' ]]; then
+            echo "::set-output name=user_email::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs | jq '.workflow_runs[0].head_commit.committer.email')"
+          else
+            echo "::set-output name=user_email::${{ github.event.pusher.email }}"
```
find how to get committer email using github env vars github actions make available to you. No need to call API.

haishulud replied:
haven't found environment variable for committer email on pull request no matter how much I've tried. Using ' git log -1 --pretty=format:%ae '

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
         id: job_conclusion
         run: |
-          echo "::set-output name=main_job_conclusion::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].conclusion')"
+          echo "::set-output name=test_job_conclusion::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].conclusion')"
+          echo "::set-output name=build_job_conclusion::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[1].conclusion')"
```
1. so long wide lines are hard to read. 
2. you have a lot of code repetition. The only thing that changes is job[X] where X is the index. So let's optimize this chunk of code.

haishulud replied:
1. done
2. in progress

haishulud replied:
2. done

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -86,60 +51,156 @@ jobs:
           secure: true
           # Body of mail message (might be a filename prefixed with file:// to read from)
-          body: Build job for ${{github.repository}} has completed. Result status ${{steps.job_conclusion.outputs.main_job_conclusion}}. Time ${{steps.calculate_time.outputs.el_hours}}h ${{steps.calculate_time.outputs.el_minutes}}m ${{steps.calculate_time.outputs.el_seconds}}s (${{steps.calculate_time.outputs.elpsd}} sec)
+          body: Build job for ${{ github.repository }} has completed. Result status for TESTS-${{ steps.job_conclusion.outputs.test_job_conclusion }}, for BUILD JOB-${{ steps.job_conclusion.outputs.build_job_conclusion }} . Time ${{ steps.calculate_time.outputs.el_hours }}h ${{ steps.calculate_time.outputs.el_minutes }}m ${{ steps.calculate_time.outputs.el_seconds }}s (${{ steps.calculate_time.ou ...
```
hard to read single line. Let's beautify the code everywhere in the pipeline and make it more readable. Keep in mind computers don't care - people do. It is the people who will read and use or maintain your code/solution. So make your solutions for humans so they can say THANK YOU - when they maintain your solution.

haishulud replied:
done

haishulud committed:
tag docker image

-tag docker image with first 7 characters of SHA1
-tag docker image with last 7 characters of SHA1

haishulud committed:
1

haishulud committed:
2

haishulud committed:
3

haishulud committed:
4

haishulud committed:
5

haishulud committed:
6

haishulud committed:
7

haishulud committed:
8

haishulud committed:
9

haishulud committed:
9

haishulud committed:
10

haishulud committed:
11

haishulud committed:
12

haishulud committed:
13

haishulud committed:
14

haishulud committed:
15

haishulud committed:
16

haishulud committed:
17

haishulud committed:
18

haishulud committed:
19

haishulud committed:
20

haishulud committed:
21

haishulud committed:
22

haishulud committed:
23

haishulud committed:
24

haishulud committed:
24

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -4,142 +4,159 @@ on:
-          echo "::set-output name=main_job_conclusion::$(curl -H "Accept: application/vnd.github+json" https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs | jq '.jobs[0].conclusion')"
+          for i in {0..1}
+          do
+          jobi=job_${i}_conclusion
+          echo "::set-output name=${jobi}::$(curl -H "Accept: application/vnd.github+json" \
+          https://api.github.com/repos/${{ github.repository }}/actions/runs/${{ github.run_id }}/jobs \
+          | jq --argjson i "$i" '.jobs[$i].conclusion')"
```
1. Pls always offset text between do/done blocks, so it's easier to read,
2. this is not optimal approach for a few reasons. Your steps (and implementation) should not depends on the structure of the pipeline. I.e your step should not even know how many jobs/steps you got. The pipeline should work the same way if you refactor the number of jobs/steps, change their namings etc. (exception is only in `needs`).

dmitry-mightydevops commented on .github/workflows/ci.yaml:
```diff
@@ -4,142 +4,159 @@ on:
+      - name: start time
+        id: start
+        run: |
+          printf -v now '%(%s)T'
```
this is a little bit cryptic. Why not to use 
```
➜ date  +%s
1661219076

```

so you get a proper value in seconds, which then you can easily operate on?
